### PR TITLE
Added more explicit iterator for keys as well as values

### PR DESCRIPTION
**Pull request by SamuraiCrow**
I've added a second iterator, unit test and documentation that allows removing of the current item and access to the key and length as well as the value.

### CODE REVIEW

**Comment by SamuraiCrow** [comment]
I found the documentation bug that killed the Mac builds on Travis.  I don't know at all why the Windows build killed the code I added even though it was used in the unit-test.

**Comment by SamuraiCrow** [comment]
Yay!  Refactoring the rehash to use the new iterator got Windows to build again!  On my Linux x64 build the latest version is only 120 bytes bigger than the old version and benchmarks comparably using GCC 10.1.

File: hashmap.h [hunk 1]
```diff
@@ -144,6 +144,18 @@ static int hashmap_iterate(const struct hashmap_s *const hashmap,
                            int (*f)(void *const context, void *const value),
                            void *const context) HASHMAP_UNUSED;
 
+/// @brief Iterate over all the heys and elements in a hashmap.
+/// @param hashmap The hashmap to insert into.
+/// @param f The function pointer to call on each element.
+/// @param context The context to pass as the first argument to f.
+/// @return If the entire hashmap was iterated then 0 is returned.
+/// Otherwise if the callback function f returned positive then non-zsero
```
Comment (magenbluten) [inline hashmap.h:154]: `non-zsero` -> `non-zero`

several issues:
- the comment overall is not very understandable
- your comment doesn't make clear that iterating stops when the callback returns a negative number

in general i would do an enum:
```
enum iterate_hint {
  ITER_CONTINUE,
  ITER_STOP,
  ITER_DELETE_CURRENT
};
```

File: hashmap.h [hunk 2]
```diff
@@ -288,6 +300,34 @@ int hashmap_iterate(const struct hashmap_s *const m,
...
+
+  /* Linear probing */
+  for (i = 0; i < m->table_size; i++)
+    p=&m->data[i];
+    if (p->in_use) {
+      r=f(context, p->key, p->key_len, p->data);
+      if (0<r) {
+        /* early return */
+        return 1;
+      }else if (0>r) {
```
Comment (magenbluten) [inline hashmap.h:323]: what happens if `r == 0` ?

File: hashmap.h [hunk 3]
```diff
@@ -288,6 +300,34 @@ int hashmap_iterate(const struct hashmap_s *const m,
...
+                    int (*f)(void *const, const char *, unsigned, void *const), void *const context) {
+  unsigned int i;
+  hashmap_element_s *p;
+  int r;
+
+  /* Linear probing */
+  for (i = 0; i < m->table_size; i++)
+    p=&m->data[i];
+    if (p->in_use) {
+      r=f(context, p->key, p->key_len, p->data);
```
Comment (magenbluten) [inline hashmap.h:319]: why not:

```
switch( f(context, &m->data[i] ) ) {
  case -1: ...
  case 0: ...
  default: ...
}
```
i pretty much dislike passing `key`, `key_len` and `data` to the callback. passing just the node is the most generic from and could even replace the existing iterator. of course it would make the internal node struct visible.

File: hashmap.h [hunk 4]
```diff
@@ -288,6 +300,34 @@ int hashmap_iterate(const struct hashmap_s *const m,
...
+      if (0<r) {
+        /* early return */
+        return 1;
+      }else if (0>r) {
+        /* Blank out the fields */
+        p->in_use = 0;
+        p->data = HASHMAP_NULL;
+        p->key = HASHMAP_NULL;
```
Comment (magenbluten) [inline hashmap.h:327]: `p->key_len` does not get reset. it is not strictly needed but i would do it for cleanliness. ( i know that this is also not done in _remove, should be there as well ).

File: hashmap.h (hunk 1 above)
Comment (SamuraiCrow) [reply hashmap.h]: Enumerations with signed values are compiler dependent for their implementations.  If the underlying data type is unsigned, it will cost more than it solves.

File: hashmap.h (hunk 3 above)
Comment (SamuraiCrow) [reply hashmap.h]: If the in_use field is set to 0 (false) it would introduce a logic error to not check it.  I will use a switch statement though.

File: hashmap.h (hunk 4 above)
Comment (SamuraiCrow) [reply hashmap.h]: Fixed.

File: hashmap.h (hunk 2 above)
Comment (SamuraiCrow) [reply hashmap.h]: 0 was the fall-through condition of -1 that would continue iteration.  Now replaced by switch statement.

File: README.md [hunk 5]
```diff
@@ -160,6 +192,7 @@ by Elliott Back. The author has applied the following further changes:
 - Changed the API to take string slices (pointer & length) instead of null
   terminated strings.
 - Did a pass to clean up the comments and function signatures.
+- Added second iterator, tests and documentation by Samuel D. Crow
```
Comment (sheredom) [inline README.md:195]: Minor minor nit thing - but can you put a period '.' at the end of the sentence?

File: hashmap.h [hunk 6]
```diff
@@ -67,6 +67,8 @@
 
 #if defined(_MSC_VER)
 #define HASHMAP_UNUSED
+#elif defined(__GNUC__)
+#define HASHMAP_UNUSED __attribute__((used))
```
Comment (sheredom) [inline hashmap.h:71]: Uh unused -> used? 🤔

File: bench/CMakeLists.txt [hunk 7]
```diff
@@ -32,3 +32,5 @@ add_executable(hashmap_bench
   ../hashmap.h
   main.c
 )
+
+target_link_libraries(hashmap_bench m)
```
Comment (sheredom) [inline bench/CMakeLists.txt:36]: Why is this needed?

Comment (SamuraiCrow) [reply bench/CMakeLists.txt:36]: No way to access the sqrt() function on GCC or Clang without the math library linked in.  I don't know whether CMake does the right thing for Visual Studio though.  I have no way to test on that platform.

File: hashmap.h (hunk 6 above)
Comment (SamuraiCrow) [reply hashmap.h:71]: I didn't invent the label name.  I don't know why it is called HASHMAP_UNUSED.

File: README.md (hunk 5 above)
Comment (SamuraiCrow) [reply README.md:195]: It's a sentence fragment but I can add one if you want one.

Comment (sheredom) [reply README.md:195]: Aye just cause the rest are non-fragments and I'm completely anal about consistency :D

File: hashmap.h (hunk 6 above)
Comment (sheredom) [reply hashmap.h:71]: Aye, given that you've now mapped it to used probably better to rename the macro!

File: bench/CMakeLists.txt (hunk 7 above)
Comment (sheredom) [reply bench/CMakeLists.txt:36]: Oh interesting - its a benchmark specific problem. I'll file a bug ta!

File: hashmap.h (hunk 6 above)
Comment (SamuraiCrow) [reply hashmap.h]: Done!

File: README.md (hunk 5 above)
Comment (SamuraiCrow) [reply README.md]: Done!

**Comment by SamuraiCrow** [comment]
The Travis task seems to be stuck.

**Comment by sheredom** [comment]
Dunno why - I just decided to merge it and hope for the best. Thanks for the change!

**Comment by SamuraiCrow** [comment]
You're welcome!
